### ISSUE TITLE
[8.13.0-rc1] broken build on RHEL-7 - undefined reference to `iszip'

### ISSUE THREAD

**Issue by remicollet** [libvips/libvips#2878]
```
/bin/sh ../libtool  --tag=CC   --mode=link gcc -std=gnu11 -std=gnu11  -O2 -g -pipe -Wall -Wp,-D_FORTIFY_SOURCE=2 -fexceptions -fstack-protector-strong --param=ssp-buffer-size=4 -grecord-gcc-switches   -m64 -mtune=generic -ftree-vectorize -no-undefined -shared -module -avoid-version -Wl,-z,relro  -o vips-magick.la -rpath /usr/lib64/vips-modules-8.13 module/vips_magick_la-magick.lo foreign/vips_magick_la-magick.lo foreign/vips_magick_la-magick2vips.lo foreign/vips_magick_la-magick6load.lo foreign/vips_magick_la-magick7load.lo foreign/vips_magick_la-vips2magick.lo libvips.la -pthread -lgmodule-2.0 -lgio-2.0 -lgobject-2.0 -lglib-2.0   -lMagickCore-6.Q16   
libtool: link: ( cd ".libs" && rm -f "vips-openslide.la" && ln -s "../vips-openslide.la" "vips-openslide.la" )
libtool: link: ( cd ".libs" && rm -f "vips-poppler.la" && ln -s "../vips-poppler.la" "vips-poppler.la" )
libtool: link: gcc -std=gnu11 -shared  -fPIC -DPIC  module/.libs/vips_magick_la-magick.o foreign/.libs/vips_magick_la-magick.o foreign/.libs/vips_magick_la-magick2vips.o foreign/.libs/vips_magick_la-magick6load.o foreign/.libs/vips_magick_la-magick7load.o foreign/.libs/vips_magick_la-vips2magick.o   -Wl,-rpath -Wl,/builddir/build/BUILD/vips-8.13.0/libvips/.libs ./.libs/libvips.so -lgmodule-2.0 -lgio-2.0 -lgobject-2.0 -lglib-2.0 -lMagickCore-6.Q16  -O2 -g -fstack-protector-strong -grecord-gcc-switches -m64 -mtune=generic -Wl,-z -Wl,relro -pthread   -pthread -Wl,-soname -Wl,vips-magick.so -o .libs/vips-magick.so
libtool: link: gcc -std=gnu11 -std=gnu11 -O2 -g -pipe -Wall -Wp,-D_FORTIFY_SOURCE=2 -fexceptions -fstack-protector-strong --param=ssp-buffer-size=4 -grecord-gcc-switches -m64 -mtune=generic -ftree-vectorize -Wl,-z -Wl,relro -Wl,-z -Wl,relro -o .libs/introspect introspect.o -I/usr/include/libgsf-1 -I/usr/include/glib-2.0 -I/usr/lib64/glib-2.0/include -I/usr/include/libxml2 -I/usr/include/orc-0.4 -pthread -I/usr/include/OpenEXR -I/usr/include/libdrm -pthread -I/usr/include/librsvg-2.0 -I/usr/include/glib-2.0 -I/usr/lib64/glib-2.0/include -I/usr/include/gdk-pixbuf-2.0 -I/usr/include/cairo -I/usr/include/libpng15 -I/usr/include/pixman-1 -I/usr/include/freetype2 -I/usr/include/uuid -I/usr/include/libdrm -D_REENTRANT -I/usr/include/cfitsio -I/usr/include/pango-1.0 -I/usr/include/cairo -I/usr/include/harfbuzz -I/usr/include/freetype2 -I/usr/include/glib-2.0 -I/usr/lib64/glib-2.0/include -I/usr/include/fribidi -I/usr/include/libpng15 -I/usr/include/uuid -I/usr/include/pixman-1 -I/usr/include/libdrm -I/usr/include/uuid -I/usr/include/freetype2 -I/usr/include/libpng15 -I/usr/include/pango-1.0 -I/usr/include/harfbuzz -I/usr/include/glib-2.0 -I/usr/lib64/glib-2.0/include -I/usr/include/fribidi -I/usr/include/libpng15 -DG_DISABLE_ASSERT -DG_DISABLE_CHECKS -pthread -I/usr/include/glib-2.0 -I/usr/lib64/glib-2.0/include -pthread -pthread  -lgirepository-1.0 ./.libs/libvips.so -lexpat -lgsf-1 -lxml2 -lfftw3 -lorc-0.4 -llcms2 -lIlmImf -lImath -lHalf -lIex -lIexMath -lIlmThread -lrsvg-2 -lgdk_pixbuf-2.0 -lmatio -lhdf5 -lz -lcfitsio -lwebpmux -lwebpdemux -lwebp -lpangocairo-1.0 -lcairo -lpangoft2-1.0 -lpango-1.0 -lfontconfig -lfreetype -ltiff -lpng15 -ljpeg -lexif -lgmodule-2.0 -lgio-2.0 -lgobject-2.0 -lglib-2.0 -lm -pthread
libtool: link: ( cd ".libs" && rm -f "vips-magick.la" && ln -s "../vips-magick.la" "vips-magick.la" )
./.libs/libvips.so: undefined reference to `iszip'
```

Trying to understand where this comes from.

P.S. build ok on RHEL-8 and RHEL-9...

**Comment by remicollet** [comment]
Removed in a9d64bea54c1fba083a9907c66e19333416a0d7e

**Comment by jcupitt** [comment]
Ooops! Thanks for catching this dumbness.

**Comment by jcupitt** [comment]
I tried in rhel7 and I see the `Already writing to another stream in archive` error too. It happens with libgsf-1.14.26.

I'll see if there's any easy workaround.

**Comment by jcupitt** [comment]
It looks like 1.14.31 (feb 2015) is the oldest libgsf that works. This issue was fixed as part of adding zip64 support.

**Comment by jcupitt** [comment]
Wow, that was painful, but I think found a simple fix. Hopefully rc2 will work.

**Comment by lovell** [comment]
I'm still investigating it, but am seeing problems with zip files produced by rc2 that were not seen with rc1, which might be related to this change.

```
$ vips dzsave input.jpg out.zip
memory: high-water mark 7.11 MB
$ unzip out.zip 
Archive:  out.zip
warning [out.zip]:  3767 extra bytes at beginning or within zipfile
  (attempting to process anyway)
file #1:  bad zipfile offset (local header sig):  3767
  (attempting to re-compensate)
out/out_files/11/0_0.jpeg:  mismatching "local" filename (out/out_files/vips-properties.xml),
         continuing with "central" filename version
out/out_files/11/0_0.jpeg:  unknown compression method
error: not enough memory for bomb detection
$ zip -T out.zip 
	zip warning: unexpected signature on disk 0 at 267579

	zip warning: archive not in correct format: out.zip
	zip warning: (try -F to attempt recovery)

zip error: Zip file structure invalid (out.zip)
```

**Comment by kleisauke** [comment]
It looks like libgsf after commit https://github.com/GNOME/libgsf/commit/e420fc52824ce5cd3d7a8bdb3eb10d848866f055 (>= 1.41.31) checks whether the output is seekable, and before that it assumes that the output is always seekable(?).

Perhaps we could reuse the `HAVE_GSF_ZIP64` definition (which checks for libgsf >= 1.41.31) and just return `FALSE` for newer libgsf versions in `gsf_output_target_seek`. For example:
```diff
--- a/libvips/foreign/dzsave.c
+++ b/libvips/foreign/dzsave.c
@@ -248,9 +248,15 @@ gsf_output_target_write( GsfOutput *output,
 static gboolean
 gsf_output_target_seek( GsfOutput *output, gsf_off_t offset, GSeekType whence )
 {
+#ifdef HAVE_GSF_ZIP64
+	/* No seek needed.
+	 */
+	return( FALSE );
+#else
 	/* This will make our parent class handle the seek.
 	 */
 	return( TRUE );
+#endif
 }
 
 static void
```
(untested on older libgsf versions, perhaps that still produces broken ZIP files)

**Comment by kleisauke** [comment]
> and before that it assumes that the output is always seekable

Confirmed, looking at the sources of libgsf 1.14.26, it's doing a seek here (assuming it will always succeed):
https://github.com/GNOME/libgsf/blob/LIBGSF_1_14_26/gsf/gsf-outfile-zip.c#L432-L434
This causes that the `zip->root->writing` is never set to `FALSE` here:
https://github.com/GNOME/libgsf/blob/LIBGSF_1_14_26/gsf/gsf-outfile-zip.c#L467
Hence, the `Already writing to another stream in archive` errors.

I just opened PR https://github.com/libvips/libvips/pull/2927 for this that fixes the broken ZIP files on both older (< 1.14.31) and newer libgsf versions.
